pywbem_mock/_baserepository.py

```diff
@@ -0,0 +1,166 @@
+pywbem client without requiring a running WBEM server.
+
+For documentation, see mocksupport.rst.
+"""
+from abc import ABCMeta, abstractmethod
+
+import six
+
+from pywbem import CIMClass, CIMQualifierDeclaration, CIMInstance
+
+
+@six.add_metaclass(ABCMeta)
+class BaseObjStore(object):
+
+    def __init__(self, case_insensitive_names, cim_object_type):
+        """
+        """
+        # TODO namespace must point to dynamic object that changes when
+        # namespaces added.
+        self._case_insensitive_names = case_insensitive_names
+        self._cim_object_type = cim_object_type
+
+    @abstractmethod
+    def exists(self, name):
+        """
+        Test if the object defined by name and namespace exists in the object
+        store
+        """
+        pass
+
+    @abstractmethod
+    def get(self, name, copy=True):
+        """
+        Return the object defined by name and namespace if it exists.
+
+        Returns:
+
+        Raises:
+            TODO: What do we return for namespace error.
+            KeyError: Object not in store
+        """
+        pass
+
+    @abstractmethod
+    def create(self, name, object):
+        """
+        """
+        pass
+
+    @abstractmethod
+    def update(self, name, object):
+        """
+        """
+        pass
+
+    @abstractmethod
+    def delete(self, name):
+        """
+        """
+        pass
+
+    def names(self):
+        """
+        Return all of the objects defined in this data store
+
+        Returns:
+            List of objects
+
+        Raises:
+
+        """
+        pass
+
+    @abstractmethod
+    def iter(self):
+        """
+        Return iterator to allow iteration through all the objects of a
+        singe namespace
+
+        Returns:
+            Iterator for a namespace
+
+        Raises:
+
+        """
+        pass
+
+
+@six.add_metaclass(ABCMeta)
+class BaseRepository(object):
+    """
+    An abstract base class for implementing CIM data store for a CIM repository.
+    This class defines the interface that is used by the
+    """
+
+    def __init__(self, repository, obj_store):
+        """
+        Construct the repository by defining the namespace with the supplied
+        default_namespace and setting the objects for the classes, instances,
+        and qualifiers
+        """
+
+        self.repository.classes = obj_store(True, CIMClass)
+        self.repository.instances = obj_store(False, CIMInstance)
+        self.repository.qualifiers = obj_store(True, CIMQualifierDeclaration)
+
+    @abstractmethod
+    def validate_namespace(self, namespace):
+        """
+        If the namespace exists in the store return. Otherwise raise
+        exception
+
+        Raises:
+
+            TODO
+        """
+        pass
+
+    @abstractmethod
+    def add_namespace(self, namespace):
+        """
+        """
+        pass
+
+    @abstractmethod
+    def remove_namespace(self, namespace):
+        """
+        Remove the namespace from this object
```

Comment:
* Better: "from this repository".
* Describe what happens when the namespace is not empty.